identity/validator_set_allegation.go

```diff
@@ -0,0 +1,344 @@
+package identity
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+
+	"github.com/Oneledger/protocol/data/balance"
+	"github.com/Oneledger/protocol/data/evidence"
+	govern "github.com/Oneledger/protocol/data/governance"
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/Oneledger/protocol/serialize"
+	"github.com/pkg/errors"
+	"github.com/tendermint/tendermint/libs/kv"
+)
+
+// TODO: Reduce power in allegation tracker for GUILTY
+// TODO: Add staking recalculation
+
+// CheckMaliciousValidators that not pass criterias and will be marked as malicious
+func (vs *ValidatorStore) CheckMaliciousValidators(es *evidence.EvidenceStore, govern *govern.Store) error {
+	vs.maliciousValidators = make(map[string]*evidence.LastValidatorHistory)
+	evidenceOptions, err := govern.GetEvidenceOptions()
+	if err != nil {
+		logger.Fatal("failed to get the evidence options")
+	}
+
+	// skip checks if does not met height criteria
+	if vs.lastHeight <= evidenceOptions.BlockVotesDiff {
+		logger.Infof("Height must be more than equal %d for check. Current: %d\n", evidenceOptions.BlockVotesDiff, vs.lastHeight)
+		return nil
+	}
+
+	if vs.lastBlockTime == nil {
+		logger.Info("Last block time is nil")
+		return nil
+	}
+
+	cv, err := es.GetCumulativeVote()
+	if err != nil {
+		return err
+	}
+
+	// fetch previous suspicious validators
+	es.IterateSuspiciousValidators(func(lvh *evidence.LastValidatorHistory) bool {
+		vs.maliciousValidators[lvh.Address.String()] = lvh
+		return false
+	})
+
+	// update found addresses with missed votes
+	for addr, votes := range cv.Addresses {
+		baddr := keys.Address{}
+		err = baddr.UnmarshalText([]byte(addr))
+		if err != nil {
+			continue
+		}
+		if votes < evidenceOptions.MinVotesRequired {
+			key := append(vs.prefix, baddr...)
+			data := vs.store.GetVersioned(vs.lastHeight-1, key)
+			if len(data) == 0 {
+				logger.Errorf("Previous state data not found for address: %s", baddr)
+				continue
+			}
+			validator := &Validator{}
+			if err := serialize.GetSerializer(serialize.JSON).Deserialize(data, validator); err != nil {
+				logger.Errorf("Validator: %s not found", baddr)
+				continue
+			}
+			validatorStatus, _ := es.GetValidatorStatus(validator.Address)
+			if validatorStatus != nil && validatorStatus.IsActive {
+				// get active from starting height as when it incomed it could be malicious, this will prevent this
+				plusDiff := validatorStatus.Height + evidenceOptions.BlockVotesDiff
+				if plusDiff > vs.lastHeight {
+					continue
+				}
+				logger.Infof("Found validator with missed required votes: %s\n", validator.Address)
+				lvh, err := es.CreateSuspiciousValidator(
+					baddr, evidence.MISSED_REQUIRED_VOTES,
+					vs.lastHeight, vs.lastBlockTime)
+				if err != nil {
+					continue
+				}
+				vs.maliciousValidators[addr] = lvh
+			}
+		}
+	}
+	return nil
+}
+
+func (vs *ValidatorStore) ExecuteAllegationTracker(ctx *ValidatorContext, activeCount int64) error {
+	if vs.lastBlockTime == nil {
```

Comment:
Should we allow unstake from a forzen validator ? @pav-kap

Reply:
@nm1918 Also here ,we should be checking if if the validator is already frozen , before removing his stake.

Reply:
you are incorrect, this is another scenario, here the steak was not deducted, here own malicious validator is created because there are not enough votes.
for the scenario where we deducted steak - so there we immediately mark that it is frozen

Reply:
I do not fully understand , what I was saying is, 
If two allegations are created against one validator.(two different request ID's)
The staking deduction should happen only once . 
If for some reason you do not want handle it here it can also be handled in action/penalization/allegation.go .
If there is still some confusion please feel free to ping me on slack

Reply:
Added for GUILTY and INNOCENT